modules/basicfuncs/list-funcs.c

```diff
@@ -70,6 +72,300 @@ tf_list_append(LogMessage *msg, gint argc, GString *argv[], GString *result)
+{
+  return list_search_parameters_parse_mode(self, argc, argv) &&
+         list_search_parameters_parse_keyword(self, argc, argv) &&
+         list_search_parameters_parse_start_index(self, argc, argv);
+}
+
+static gboolean
+list_search_parameters_init(ListSearchParameters *self, gint *argc, GString **argv[])
+{
+  return list_search_parameters_parse_parameters(self, argc, argv) &&
+         list_search_pattern_prepare(self->pattern, self->mode);
+}
+
+static void
+list_search_parameters_free(ListSearchParameters *self)
+{
+  if (self->pattern)
+    list_search_pattern_free(self->pattern);
+  g_free(self);
+}
+
+static ListSearchParameters *
+list_search_parameters_new(void)
+{
+  ListSearchParameters *self = g_new0(ListSearchParameters, 1);
+  self->pattern = list_search_pattern_new();
+
+  return self;
+}
+
+static gboolean
+_list_search_match_pcre(const GString *string, ListSearchPattern *pattern)
+{
+  gint rc;
+  gint num_matches;
+
+  if (pcre_fullinfo(pattern->pcre, pattern->pcre_extra, PCRE_INFO_CAPTURECOUNT, &num_matches) < 0)
+    g_assert_not_reached();
+  if (num_matches > RE_MAX_MATCHES)
+    num_matches = RE_MAX_MATCHES;
+
+  gsize matches_size = 3 * (num_matches + 1);
+  gint *matches = g_alloca(matches_size * sizeof(gint));
+
+  rc = pcre_exec(pattern->pcre, pattern->pcre_extra, string->str, string->len, 0, 0, matches, matches_size);
+  if (rc == PCRE_ERROR_NOMATCH)
+    {
+      return FALSE;
+    }
+  if (rc < 0)
+    {
+      msg_error("Error while matching pcrep", evt_tag_int("error_code", rc));
+      return FALSE;
+    }
+  if (rc == 0)
+    {
+      msg_error("Error while storing matching substrings");
+      return FALSE;
+    }
+  return TRUE;
+}
+
+static gboolean
+_list_search_match(const GString *string, ListSearchParameters *parameters)
+{
+  ListSearchPattern *pattern = parameters->pattern;
+
+  switch (parameters->mode)
+    {
+    case LSM_LITERAL:
+      return (g_string_equal(string, pattern->keyword));
+    case LSM_PREFIX:
+      return (strncmp(string->str, pattern->keyword->str, pattern->keyword->len) == 0);
+    case LSM_SUBSTRING:
+      return (strstr(string->str, pattern->keyword->str) != NULL);
+    case LSM_GLOB:
+      return (g_pattern_match(pattern->glob, string->len, string->str, pattern->keyword_reversed->str));
+    case LSM_PCRE:
+      return (_list_search_match_pcre(string, pattern));
+    default:
+      g_assert_not_reached();
+    }
+}
+
+static void
+_list_search(gint argc, GString *argv[], GString *result, ListSearchParameters *parameters)
+{
+  if (argc < 1)
+    return;
+
+  ListScanner scanner;
+  gint index = -1;
+
+  list_scanner_init(&scanner);
+  list_scanner_input_gstring_array(&scanner, argc, argv);
+
+  while (list_scanner_scan_next(&scanner))
+    {
+      if (++index < parameters->start_index)
+        continue;
+
+      if (_list_search_match(list_scanner_get_current_value_as_gstring(&scanner), parameters))
```

Comment:
I am not fun of the `get_as_type` like functions. In this case I think the current `list_scanner_get_current_value` could return with `GString`, and the users could be updated accordingly.

Reply:
I went the other way, and converted `match()` to handle `char *` and `gsize` instead of `GString*`.